core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java

```diff
@@ -34,72 +33,93 @@ private UfsSyncUtils() {} // prevent instantiation
       }
+      return syncPlan;
+    }
+
+    // One of the metadata or content is not in sync
+    if (inode.isDirectory() && (containsMountPoint || ufsIsDir)) {
+      // Instead of deleting and then loading metadata to update, try to update directly
+      // - mount points (or paths with mount point descendants) should not be deleted
+      // - directory permissions can be updated without removing the inode
+      if (inode.getParentId() != InodeTree.NO_PARENT) {
+        // Only update the inode if it is not the root directory. The root directory is a special
+        // case, since it is expected to be owned by the process that starts the master, and not
+        // the owner on UFS.
+        syncPlan.setUpdateDirectory();
+      }
+      syncPlan.setSyncChildren();
+      return syncPlan;
+    }
+
+    // One of metadata or content is not in sync and it is a file
+    if (!isContentSynced) {
+      // update inode, by deleting and then optionally loading metadata
+      syncPlan.setDelete();
+      if (ufsExists) {
+        // UFS exists, so load metadata later.
+        syncPlan.setLoadMetadata();
+      }
+    } else {
+      syncPlan.setUpdateMetadata();
     }
     return syncPlan;
   }
 
   /**
-   * Returns true if the given inode is synced with the ufs status. This is a single inode check,
-   * so for directory inodes, this does not consider the children inodes.
+   * Returns true if the given inode's content is synced with the ufs status. This is a single inode
+   * check, so for directory inodes, this does not consider the children inodes.
    *
    * @param inode the inode to check for sync
    * @param ufsFingerprint the ufs fingerprint to check for the sync
    * @return true of the inode is synced with the ufs status
    */
-  public static boolean inodeUfsIsSynced(Inode inode, String ufsFingerprint) {
+  public static boolean inodeUfsIsContentSynced(Inode inode, Fingerprint ufsFingerprint) {
     boolean isSyncedUnpersisted =
-        !inode.isPersisted() && Constants.INVALID_UFS_FINGERPRINT.equals(ufsFingerprint);
-
+        !inode.isPersisted() && !ufsFingerprint.isValid();
     boolean isSyncedPersisted;
+
     if (inode instanceof InodeFile) {
       // check the file fingerprint.
       InodeFile inodeFile = (InodeFile) inode;
+      Fingerprint cachedFingerprint =  Fingerprint.parse(inodeFile.getUfsFingerprint());
       isSyncedPersisted = inodeFile.isPersisted()
-          && inodeFile.getUfsFingerprint().equals(ufsFingerprint)
-          && !inodeFile.getUfsFingerprint().equals(Constants.INVALID_UFS_FINGERPRINT);
+          && cachedFingerprint.matchContent(ufsFingerprint)
+          && cachedFingerprint.isValid();
     } else {
-      isSyncedPersisted = inode.isPersisted() && inode.getUfsFingerprint().equals(ufsFingerprint);
+      Fingerprint cachedFingerprint =  Fingerprint.parse(inode.getUfsFingerprint());
+      isSyncedPersisted = inode.isPersisted() && cachedFingerprint.matchContent(ufsFingerprint);
```

Comment:
Do you think we can combine the dir and file paths to be the same, now that we have better fingerprint functionality?

Reply:
done